patches/sqlite/0002-sqlite-complete-early-exit.patch

```diff
@@ -0,0 +1,61 @@
+--- sqlite-src-pristine/src/complete.c	2023-11-01 22:31:37
++++ sqlite-src-modified/src/complete.c	2024-03-14 12:51:32
+@@ -99,14 +99,21 @@
+ **
+ ** If we compile with SQLITE_OMIT_TRIGGER, all of the computation needed
+ ** to recognize the end of a trigger can be omitted.  All we have to do
+ ** is look for a semicolon that is not part of an string or comment.
+ */
+-int sqlite3_complete(const char *zSql){
++int sqlite3_complete(const char *zSql) {
++    return sqlite3_complete_length(zSql, 0) != 0;
++}
++
++int sqlite3_complete_length(const char *zSql, int firstOnly){
+   u8 state = 0;   /* Current state, using numbers defined in header comment */
+   u8 token;       /* Value of the next token */
+
++  /* Store the original pointer for later index calculations */
++  const char *initialZSql = zSql;
++
+ #ifndef SQLITE_OMIT_TRIGGER
+   /* A complex statement machine used to detect the end of a CREATE TRIGGER
+   ** statement.  This is the normal case.
+   */
+   static const u8 trans[8][8] = {
+@@ -254,12 +261,16 @@
+         break;
+       }
+     }
+     state = trans[state][token];
+     zSql++;
++
++    // Once we have found a single complete statement, return its length
++    if (firstOnly && state==1) return zSql - initialZSql;
+   }
+-  return state==1;
++  // Otherwise only return the length if the entire input is valid
++  return state==1 ? zSql - initialZSql : 0;
```

Comment:
I suspect your tests don't cover this line because they always take the earlier return. In fact, I suppose when `firstOnly` is true, then this line can only be hit when `state` is 0?

I'm not sure there's any action to take here, it would be weird to write tests for code in sqlite that we don't use. But just checking that this matches your understanding...

Reply:
Yeah we don't use `sqlite3_complete` in its original version anywhere. I have been manually running the SQLite suite on the code with this patch though and it's passing those tests.